# [CPSC 222](https://github.com/GonzagaCPSC222) Intro to Data Science
[Gonzaga University](https://www.gonzaga.edu/)

[Gina Sprint](http://cs.gonzaga.edu/faculty/sprint/)

# Nearest Neighbor Classification
What are our learning objectives for this lesson?
* Learn about the kNN classification algorithm

Content used in this lesson is based upon information in the following sources:
* Dr. Shawn Bowers' Data Mining notes

## Warm-up Task(s)
1. Finish your MA12 w/your partner
1. Ask me any DA6 questions you might have
1. Open VS Code or Jupyter Lab in a new folder called ClassificationFun
    * Create a new notebook called kNN.ipynb
    * Create a `DataFrame` for the following data:
```
height(cm),weight(kg),t-shirt size
158,58,M
163,61,M
165,61,L
168,66,L
```

## Today
* Announcements
    * MA13 notecard quiz next class!
    * DA6 is due tonight. Come to my office hours or TA Alicia's office hours to check your step 1s
    * Work on MA13 and your project this weekend
    * Tonight: software copyright & AI talk. PACCAR 107 at 6pm
* MA12/DA6 work time
    * MA12 solutions
* Starting our machine learning unit with the kNN Lab 🏠
* IQ8 last ~15 mins of class

## Warm-up Task(s) 4/15
1. Pull out your MA13 note sheet and the kNN Lab
1. Open ClassificationFun/kNN.ipynb
1. That's it!

## Today 4/15
* Announcements
    * Nice job on IQ8, let's go over it
    * IQ9 next class on U6/DA6 (emphasis on confidence intervals and hypothesis testing)
        * As long as you don't have a bunch of notes/writing on your stats cheat sheet, you can use it on the quiz!
        * If you need to print a fresh stats cheat sheet, you can find it in our [Google Drive folder](https://drive.google.com/drive/u/0/folders/1mzkIiJwJM_R4Vjf6YIdKqbwVtCg66Ncl)
    * DA7 is posted. Please read through it before next class
    * Work on your project (Cleaning, EDA, and at least 1 hypothesis test)
        * Check-in due sometime after we come back from Easter break (in class on 4/29 at the latest)
        * Bonus points for demoing early (week of 4/22-4/25) during office hours
    * Reminder about game dev talk Thursday night 6pm Bollier 120
* MA13
* Finish kNN Lab
* ClassificationFun (finish kNN lab + kNN code in kNN.ipynb)
* Go over MA12 solutions

## Nearest Neighbor Classification
Nearest neighbor classification is typically used when attributes are continuous
* Can be modified for categorical data ...

Basic approach:
* Given an instance $i$ with $n - 1$ attributes (where the $n^{th}$ is class label)
* Find the "closest" instance $j$ to $i$ on the $n - 1$ attributes
* Use $j$'s class as the prediction for $i$

Example from [Bramer](https://www.amazon.com/Principles-Mining-Undergraduate-Computer-Science/dp/1447148835):
Given the data set

|a |b |c |d |e |f |Class|
|-|-|-|-|-|-|-|
|yes |no |no |6.4 |8.3 |low |negative|
|yes |yes |yes |18.2 |4.7 |high |positive|

What should this instance's classification be?

|yes |no |no |6.6 |8.0 |low |???|
|-|-|-|-|-|-|-|

Usually it isn't this easy!

## k Nearest Neighbors
Find the k nearest neighbors ...
* Usually find the k closest neighbors (instead of just closest)
* Then pick classification from among the top k

What are good values for the number of neighbors k?
* Often done experimentally (with a test set)
* Start with k = 1, determine "error rate" (more later)
* Repeat incrementing k
* Pick k with smallest (minimum) error rate
     * Often, larger the data (training) set, the larger the k

Lets say we found the k nearest neighbors for an instance ...

Q: What are ways we could pick the class?
* Most frequent occurring class
* Weighted "average" (based on the relative closest of the k)

Note: can use k-NN for regression if, e.g., return the mean of the label values

## Distance Functions
k-NN works by calculating distances between instances
* Many possible ways to do this ... generalized through "distance measures"
* For two points $x$ and $y$, the distance between them is given by $dist(x,y)$

Properties of distance measures (metrics)
1. $\forall x$, $dist(x,x) = 0$
    * The distance of any point x from itself is zero
2. $\forall xy$, $dist(x,y) = dist(y,x)$
    * Symmetry
3. $\forall xyz$, $dist(x,y) \leq dist(x,z) + dist(z,y)$
    * Triangle equality
    * "Shortest distance between any two points is a straight line"

Euclidean Distance is most often used: Given an instance, treat it as a "vector" in n space
* Use Pythagoras' Theorem to find distance between them

For example:
* Given two points (i.e., rows) with $n = 2$: $(x_1, y_1)$ and $(x_2, y_2)$
* The length (i.e., distance) of the straight line joining the points is:

$$\sqrt{(x_1 - x_2)^{2} + (y_1 - y_2)^{2}}$$

* Euclidean $n$-space
    * For rows A = $(a_1, a_2,..., a_n)$ and $B = (b_1, b_2,..., b_n)$ with $n$ attributes
$$\sqrt{(a_1 - b_1)^{2} + (a_2 - b_2)^{2} +...+ (a_n - b_n)^{2}}$$
        * Which is:
$$\sqrt{\sum_{i=1}^{n}(a_i - b_i)^{2}}$$

Other examples are described in the book (e.g., Manhattan "city block" distance)

Q: Do you see any possible issues with Euclidean distance?
* Larger values tend to dominate smaller ones
* Can degrade into a few attributes driving the distances
    * e.g., [Mileage=18,457, Doors=2, Age=12]
    
## Normalization
One solution is to scale all values between 0 and 1 ("min-max" normalization)
* Use the formula: `(x - min(xs)) / ((max(xs) - min(xs)) * 1.0)`

Q: How can we deal with categorical values?
* $dist(v_1, v_2) = 0$ if values $v_1 = v_2$
    * Same values
* For nominal values, $dist(v_1, v_2) = 1$ if $v_1 \neq v_2$
    * Different values
* For ordinal values, assign to 1 or use the "distance"

Q: What do we do about missing values?
* Don't have missing values
    * Clean the data first
* Be conservative
    * Assuming normalized
    * If only one value missing:
        * Assume the maximum possible distance
        * If nominal use the maximum distance (i.e., 1)
        * If ordinal use either 1 or furthest distance from known value
    * otherwise, if both values missing, use the maximum distance (e.g., 1)

Distance-based metrics imply equal weighting of attributes
* Sometimes can perform better with attribute weights (i.e., some attributes worth more)
* "Feature reduction" (not using certain attributes) can also help with redundant or "noisy" attributes

## Basic k-NN Algorithm
```
Input: list of rows, no of atts (n where nth is label), instance to classify, k
def kNN_classifier(training_set, n, instance, k):
    row_distances = []
    for row in training_set:
        d = distance(row, instance, n - 1)
        row_distances.append([d, row])
    top_k_rows = get_top_k(row_distances, k)
    label = select_class_label(top_k_rows)
    return label
```

## kNN Example
Example adapted from [this kNN example](https://people.revoledu.com/kardi/tutorial/KNN/KNN_Numerical-example.html)

Suppose we have the following dataset that has two attributes (acid durability and strength) and a class attribute (whether a special paper tissue is good or not):

|Acid durability (seconds)|Strength (kg/square meter)|Classification|
|-|-|-|
|7|7|Bad|
|7|4|Bad|
|3|4|Good|
|1|4| Good|

Now the factory produces a new paper tissue with acid durability = 3 seconds and strength = 7 kg/square meter. Can we predict what the classification of this new tissue is? Use kNN with $k$ = 3. 

### Make a Prediction Manually
Steps:
1. Normalize
1. Compute distance of each training instance to the test instance
1. Determine the majority classification of the $k$ closest instances... this is your prediction for the test instance

After normalization:

|Acid durability (seconds)|Strength (kg/square meter)|Classification|
|-|-|-|
|1|1|Bad|
|1|0|Bad|
|0.33|0|Good|
|0|0| Good|

Test instance normalization: 0.33, 1

Distances:

|Acid durability (seconds)|Strength (kg/square meter)|Classification|Distance|
|-|-|-|-|
|1|1|Bad|0.66|
|1|0|Bad|1.203|
|0.33|0|Good|1.0|
|0|0| Good|1.05|

Work:
* $\sqrt{(1-0.33)^2 + (1-1)^2} = 0.66$
* $\sqrt{(1-0.33)^2 + (0-1)^2} = 1.203$
* $\sqrt{(0.33-0.33)^2 + (0-1)^2} = 1.0$
* $\sqrt{(0-0.33)^2 + (0-1)^2} = 1.05$

Majority classification: 
1 Bad (0.66) and 2 Goods (1.0 an 1.05) => Good!

### Make a Prediction with Scikit-Learn
Steps:
1. Load data
1. Normalize
1. Train kNN classifier with training set
1. Test kNN classifier on test instance

In [1]:
# load data
import pandas as pd

data = [[7, 7, "Bad"], [7, 4, "Bad"], [3, 4, "Good"], [1, 4, "Good"]]
df = pd.DataFrame(data, columns=["Acid durability (seconds)", "Strength (kg/square meter)", "Classification"])

print(df)

   Acid durability (seconds)  Strength (kg/square meter) Classification
0                          7                           7            Bad
1                          7                           4            Bad
2                          3                           4           Good
3                          1                           4           Good


In [7]:
# normalize
from sklearn.preprocessing import MinMaxScaler

X_train = df.drop("Classification", axis=1)
y_train = df["Classification"]

scaler = MinMaxScaler()
scaler.fit(X_train)
print(scaler.data_min_)
print(scaler.data_max_)

X_train_normalized = scaler.transform(X_train)
print(X_train_normalized)

[1. 4.]
[7. 7.]
[[1.         1.        ]
 [1.         0.        ]
 [0.33333333 0.        ]
 [0.         0.        ]]


In [8]:
# train
from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier(n_neighbors=3)
neigh.fit(X_train_normalized, y_train)

# test
X_test = pd.Series([3, 7], index=df.columns.drop("Classification"))
X_test = scaler.transform([X_test])
y_test_prediction = neigh.predict(X_test)
print(y_test_prediction)

['Good']


### Some Notes
Q: What happens if there are ties in the top-k distances (get_top_k)? E.g., which are top 3 in: [[.28,$r_1$],[.33,$r_2$],[.33,$r_3$],[.33,$r_4$],[.37,$r_5$]]?
* Different options ... e.g.:
    * Randomly select from ties
    * Do top-k distances (instead of instances)
    * Ignore ties (in case above, just use $r_1$ and $r_2$)

Nearest doesn't imply near
* top-k instances might not be that close to the instance being classified
* Especially true as the number of attributes ("dimensions") increases
    * An example of the "curse of dimensionality"
* Again, have to use common sense and an understanding of the dataset

### Efficiency issues
Q: Is k-NN efficient? Can you find any efficiency issues?
* Given a training set with $D$ instances and $k = 1$
* $O(D)$ comparisons needed to classify a given instance

Q: Can you think of any ways to improve the efficiency?
1. Use search trees
    * Presort and arrange instances into a search tree
    * Can reduce comparisons to $O(log D)$
2. Check each training instance in parallel
    * Gives $O(1)$ comparisons
3. Editing/Pruning
    * Filter or remove training tuples that prove useless
    * Reduces size of $D$